mrmustard/math/lattice/strategies/beamsplitter.py

```diff
@@ -202,3 +202,74 @@ def beamsplitter_schwinger(shape, theta, phi, max_N=None):
             for j in range(max(0, N + 1 - c1), min(N + 1, c2)):
                 U[N - i, i, N - j, j] = block[i, j]
     return U
+
+
+def sector_idx(N: int, shape: tuple):
+    """The action of a BSgate breaks down into N-dim unitaries acting
+    on each total photon number subspace, because the BS commutes with the
+    total photon number operator.
+
+    e.g. we have a two-mode ket of initial shape (4,4) (i.e. 0 to 3 photons in each mode)
+    The 3x3 BS unitary on the 2-photon subspace acts on indices [2,5,8] of
+    the flattened ket array.
+    Here is the ordered basis vectors |m,n> by total photons m+n for the example:
+
+    flat index  |m,n>
+    0            0,0  |-- 0 photons
+
+    1            1,0  |-- 1 photons
+    4            0,1  |
+
+    2            2,0  |-- 2 photons  <-- 2-photon subspace at indices [2,5,8]
+    5            1,1  |
+    8            0,2  |
+
+    3            3,0  |-- 3 photons
+    6            2,1  |
+    9            1,2  |
+    12           0,3  |
+
+    7            3,1  |-- 4 photons
+    10           2,2  |
+    13           1,3  |
+
+    11           3,2  |-- 5 photons
+    14           2,3  |
+
+    15           3,3  |-- 6 photons
+
+    The left column is the flattened order. This function returns the indices in
+    left column for the N-th block. E.g. sector_idx(3, (4,4)) is [3,6,9,12].
+    """
+    return [
+        np.ravel_multi_index((i, N - i), shape) for i in range(N + 1) if max(i, N - i) < max(shape)
```

Comment:
Is the `if max(i, N - i) < max(shape)` intended to handle out of bound indices? If so, could it be handled by the `mode` param in `np.ravel_multi_index` instead?

Reply:
Good question, I didn't think of `mode`. I had a look, but it doesn't seem any of the modes allows you to skip some indices?

Reply:
Oh I see, we could use `mode='clip'` and then filter out the zeros

Reply:
hmm I tried, but it clips the upper boundary to the max index, so not zero. I'm afraid `mode` doesn't help.